benchmarks/be_eval.erl

```diff
@@ -0,0 +1,231 @@
+-module(be_eval).
+
+-behaviour(gen_server).
+
+-include("be_eval.hrl").
+
+%% API
+-export([
+  start_link/1,
+  stop/1,
+  run/6,
+  run/8
+]).
+
+%% gen_server callbacks
+-export([init/1, handle_call/3, handle_cast/2, handle_info/2, terminate/2,
+  code_change/3]).
+
+-define(SERVER, ?MODULE).
+
+-record(be_eval_state, {
+  id,
+  eval_count
+}).
+
+%%%===================================================================
+%%% API
+%%%===================================================================
+
+%% @doc Spawns the server and registers the local name (unique)
+-spec(start_link(Id :: atom()) ->
+  {ok, Pid :: pid()} | ignore | {error, Reason :: term()}).
+start_link(Id) ->
+  gen_server:start_link({local, Id}, ?MODULE, [Id], []).
+
+%% @doc Stops the server
+stop(ServerRef) ->
+  gen_server:stop(ServerRef).
+
+%% @doc Loads a BE and evaluates events
+run(ServerRef, Info, BetreeFile, EventsFile, EventEvalFunc, StatsFunc) ->
+  run(ServerRef, Info, BetreeFile, EventsFile, EventEvalFunc, _EventEvalOutputFile = undefined,
+    StatsFunc, _StatsOutputFile = undefined).
+
+run(ServerRef, Info, BetreeFile, EventsFile, EventEvalFunc, EventEvalOutputFile, StatsFunc, StatsOutputFile) ->
+  gen_server:call(ServerRef,
+    {run, Info, BetreeFile, EventsFile, EventEvalFunc, EventEvalOutputFile, StatsFunc, StatsOutputFile},
+    infinity).
+
+%%%===================================================================
+%%% gen_server callbacks
+%%%===================================================================
+
+%% @private
+%% @doc Initializes the server
+-spec(init(Args :: term()) ->
+  {ok, State :: #be_eval_state{}} | {ok, State :: #be_eval_state{}, timeout() | hibernate} |
+  {stop, Reason :: term()} | ignore).
+init([Id]) ->
+  {ok, #be_eval_state{id = Id, eval_count = 0}}.
+
+%% @private
+%% @doc Handling call messages
+-spec(handle_call(Request :: term(), From :: {pid(), Tag :: term()},
+    State :: #be_eval_state{}) ->
+  {reply, Reply :: term(), NewState :: #be_eval_state{}} |
+  {reply, Reply :: term(), NewState :: #be_eval_state{}, timeout() | hibernate} |
+  {noreply, NewState :: #be_eval_state{}} |
+  {noreply, NewState :: #be_eval_state{}, timeout() | hibernate} |
+  {stop, Reason :: term(), Reply :: term(), NewState :: #be_eval_state{}} |
+  {stop, Reason :: term(), NewState :: #be_eval_state{}}).
+handle_call({run, Info, BetreeFile, EventsFile, EventEvalFunc, EventEvalOutputFile, StatsFunc, StatsOutputFile},
+    _From, State = #be_eval_state{
+    id = Id, eval_count = EvalCount
+  }) ->
+  case be_bm_utils:file_exists(EventsFile) of
+    false ->
+      {reply, {error, {file_does_not_exist, EventsFile}}, State};
+    true ->
+      EvalCount1 = EvalCount + 1,
+      StrId = atom_to_list(Id),
+      StrCount = integer_to_list(EvalCount1),
+      LoaderId = list_to_atom(StrId ++ "_" ++ StrCount ++ "_loader"),
```

Comment:
we can have potential memory leak here with creating atoms with counter. By default maximum number of atoms is 1,048,576

Reply:
Not so much of a memory leak, but a potential atom leak.  Do the spawned processes need to have an atom name?

Reply:
Atoms leak addressed.